pkg/provisioner/ironic/ironic.go

```diff
@@ -1,62 +1,313 @@
+		ironicNode, err = nodes.Get(p.client, p.host.Name).Extract()
+		switch err.(type) {
+		case nil:
+			p.log.Info("found existing node by name")
+			// Store the ID so other methods can assume it is set and
+			// so we can find the node using that value next time.
+			p.status.ID = ironicNode.UUID
+			dirty = true
+			p.log.Info("setting provisioning id", "ID", p.status.ID)
+		case gophercloud.ErrDefault404:
+			p.log.Info("did not find existing node by name")
+			ironicNode = nil
+		default:
+			return false, errors.Wrap(err,
+				fmt.Sprintf("failed to find node by name %s", p.host.Name))
+		}
+	}
+
+	// If we have not found a node yet, we need to create one
+	if ironicNode == nil {
+		p.log.Info("registering host in ironic")
+
+		driverInfo := p.bmcAccess.DriverInfo(p.bmcCreds)
+		// FIXME(dhellmann): The names of the images are tied
+		// to the version of ironic we are using and are
+		// likely to change.
+		//
+		// FIXME(dhellmann): We need to get our IP on the
+		// provisioning network from somewhere.
+		driverInfo["deploy_kernel"] = "http://172.22.0.1/images/tinyipa-stable-rocky.vmlinuz"
+		driverInfo["deploy_ramdisk"] = "http://172.22.0.1/images/tinyipa-stable-rocky.gz"
+
+		ironicNode, err = nodes.Create(
+			p.client,
+			nodes.CreateOpts{
+				Driver:        "ipmi",
+				BootInterface: "pxe",
+				Name:          p.host.Name,
+				DriverInfo:    driverInfo,
+			}).Extract()
+		if err != nil {
+			return false, errors.Wrap(err, "failed to register host in ironic")
+		}
+
+		// Store the ID so other methods can assume it is set and so
+		// we can find the node again later.
+		p.status.ID = ironicNode.UUID
 		dirty = true
+		p.log.Info("setting provisioning id", "ID", p.status.ID)
+
+		// If we know the MAC, create a port. Otherwise we will have
+		// to do this after we run the introspection step.
+		if p.host.Spec.BootMACAddress != "" {
+			enable := true
+			p.log.Info("creating port for node in ironic", "MAC",
+				p.host.Spec.BootMACAddress)
+			_, err := ports.Create(
+				p.client,
+				ports.CreateOpts{
+					NodeUUID:   ironicNode.UUID,
+					Address:    p.host.Spec.BootMACAddress,
+					PXEEnabled: &enable,
+				}).Extract()
+			if err != nil {
+				return false, errors.Wrap(err, "failed to create port in ironic")
+			}
+		}
+
+		// FIXME(dhellmann): We should test whether the incoming
+		// values match the values on the node in ironic before
+		// updating them. For now, using "dirty" flag as an indicator
+		// that we created the node so we need to set these values.
+		p.log.Info("updating host settings in ironic")
+		_, err = nodes.Update(
+			p.client,
+			ironicNode.UUID,
+			nodes.UpdateOpts{
+				nodes.UpdateOperation{
+					Op:    nodes.AddOp,
+					Path:  "/instance_info/image_source",
+					Value: p.instanceImageSource,
+				},
+				nodes.UpdateOperation{
+					Op:    nodes.AddOp,
+					Path:  "/instance_info/image_checksum",
+					Value: p.instanceImageChecksum,
+				},
```

Comment:
You need to support `/instance_info/root_gb` (I don't remember why we need it for whole disk images, but can investigate) and `properties/root_device` (optional, but much recommended for multi-disk machines).

Reply:
What values should be used for those?

Reply:
Good question. I went through the sources yesteday, and I'm pretty sure this value should NOT be required for whole-disk images. I'm going to report and fix it today, but meanwhile you can put any positive integer. Say, 10.

Reply:
Ehhhm, I was talking about `root_gb`. The `root_device` should either come from a user or be detected using introspection data. We don't have a good default.

Reply:
`root_gb`: https://storyboard.openstack.org/#!/story/2005165

Reply:
Done